gnomad_hail/resources/resource_utils.py

```diff
@@ -9,114 +9,221 @@
+            import_func=import_func,
+            expected_file_extensions=['.mt']
         )
 
-    def mt(self) -> hl.MatrixTable:
+    def mt(self, force_import: bool = False) -> hl.MatrixTable:
         """
         Read and return the Hail MatrixTable resource
 
         :return: Hail MatrixTable resource
         """
-        return hl.read_matrix_table(self.path)
+        if self.path is None or force_import:
+            return self.import_func(**self.import_sources)
+        else:
+            return hl.read_matrix_table(self.path)
+
+    def import_resource(self, overwrite: bool = False, **kwargs):
+        """
+        Imports the MatrixTable resource using its import_func and writes it in its path.
+
+        :param overwrite: If set, existing file(s) will be overwritten
+        :param kwargs: Any other parameters to be passed to hl.MatrixTable.write
+        :return: Nothing
+        """
+        self.import_func(**self.import_sources).write(self.path, overwrite=overwrite, **kwargs)
 
 
 class PedigreeResource(BaseResource):
     """
     A pedigree resource
+
+    :param path: The Pedigree path (typically ending in .fam or .ped)
+    :param import_sources: Any sources that are required for the import and need to be kept track of and/or passed to the import_func (e.g. .vcf path for an imported VCF)
+    :param import_func: A function used to import the MatrixTable. `import_func` will be passed the `import_sources` dictionary as kwargs.
+    :param quant_pheno: If ``True``, phenotype is interpreted as quantitative.
+    :param delimiter: Field delimiter regex.
+    :param missing: The string used to denote missing values. For case-control, 0, -9, and non-numeric are also treated as missing.
     """
 
-    def __init__(self, path: str, import_sources: Optional[Dict[str, Any]] = None, ):
+    def __init__(
+            self,
+            path: Optional[str] = None,
+            import_sources: Optional[Dict[str, Any]] = None,
+            import_func: Optional[Callable[..., hl.MatrixTable]] = None,
+            quant_pheno: bool = False,
+            delimiter: str = r"\\s+",
+            missing: str = 'NA'
+    ):
         super().__init__(
             path=path,
             import_sources=import_sources,
-            expected_file_extension='.fam'
+            import_func=import_func,
+            expected_file_extensions=['.fam', '.ped']
         )
 
-    def ht(self, delimiter=r"\\s+") -> hl.Table:
+        self.quant_pheno = quant_pheno
+        self.delimiter = delimiter
+        self.missing = missing
+
+    def ht(self) -> hl.Table:
         """
         Reads the pedigree into a family HT using hl.import_fam().
 
-        :param delimiter: Delimiter used in the ped file
         :return: Family table
         """
-        return hl.import_fam(self.path, delimiter=delimiter)
+        return hl.import_fam(self.path, quant_pheno=self.quant_pheno, delimiter=self.delimiter, missing=self.missing)
```

Comment:
The `ht` method on this class seems out of place. Other resource classes have _one_ method to get the Hail object they're associated with (`ht` for TableResource, `mt` for MatrixTableResource, `bm` for BlockMatrixResource). Why does PedigreeResource have another method in addition to `pedigree`? And it requires adding to the class two more attributes (`quant_pheno` and `missing`) that are ignored by the primary method `pedigree`.

Reply:
Pedigree information is weird like this in that sometimes we use it in a python context (hence the `.pedigree()` function and sometimes in a hail annotation context that requires a `ht`. In my opinion bundling everything about the resource within the class is the option that makes the most sense.

Reply:
I kind of want to move these to `import_sources` (though the "sources" name doesn't quite fit) since they're related to importing data from a non-Hail native file format.

Reply:
We sort of do this with importing the dbsnp header file? I know it is an actual file so a source but also an extra parameter in the hail function?

Reply:
Maybe we should rename `import_sources` to something more generic like `import_info`?

Reply:
That works for me

Reply:
we could rename `import_sources` to `import_info`, that said these aren't about importing but are pieces of information about the resource which only pertain to this specific resource type.

Reply:
actually, I was thinking of `quant_pheno` -- I agree that the rest _is_ about importing.

Reply:
They're arguments to **import**_fam. There's no requirement or assumption on what keys should be in `import_sources`/`import_info` for different resource types or even individual resources. I think the fact that `import_info` is per-individual resource is a better fit here since different individual pedigree resources are used differently (with `.ht()` or with `.pedigree()`).

Reply:
so here I think the weird bit is - as Nick pointed out - the fact that we store this information in plain text as there is no Hail-native format (hail's `Pedigree.write` just writes a text file). For this reason, `import` here as a bit of a different meaning in that it'd really be a function that converts some other format into...well a text format (fam file). 
Now when we use the resource, the fam file is read and unfortunately at that point additional args are needed. I truly believe that storing these with the resource will provide the best user experience and don't see these args as `import_info`.